script/proto/proto.lua

```diff
@@ -138,12 +143,17 @@ function m.doMethod(proto)
             end
             await.close('proto:' .. proto.id)
             if ok then
-                m.response(proto.id, res)
+                if completed or not m.errorOnCancel[method] then
+                    m.response(proto.id, res)
+                else
+                    m.responseErr(proto.id, define.ErrorCodes.RequestCancelled, res)
```

Comment:
A request may be interrupted for other reasons, such as ·`ContentModified: number = -32801;`, so it cannot be simply assumed that the request is canceled.

You can consider proactively setting the reason for closing the task before closing the task.
For https://github.com/sumneko/lua-language-server/blob/1f0a2d0e9283a4cb7f7b3fc72258eb1c5ba4e5dd/script/provider/provider.lua#L705 this situation may be troublesome, because I don’t want to deal with `proto` in `await` .The transaction may only be notified through the callback function

Reply:
https://github.com/sumneko/lua-language-server/commit/eb939ecccd11c6c628fc478f6e4550ab9c66a95c

```lua
await.setID('update:' .. uri, function ()
    proto.close(id, REASON)
end)
```

Maybe like this

Reply:
This issue is not related to this PR, maybe we can treat it as a separate issue

Reply:
I made the `RequestCancelled` error be an opt-in by method instead globally (so it won't apply to anything but the `semanticTokens` methods right now, or others that pass `true` as a third parameter to `proto.on`), but I see the point you're making...technically even the `semanticTokens` could fail to complete otherwise. At one point I had tried storing a `canceled` map with `proto.id` as the key, but it wasn't being picked up. I can try playing around some more today...

The reason I included that bit in this PR is because it fixed the issue of semantic highlighting going away while typing. The client has to get `RequestCancelled` instead of a nil response for the `semanticTokens` endpoints or it'll apparently clear out all the current tokens.

Reply:
I think this is a bug by https://github.com/sumneko/lua-language-server/commit/e5a094cfc9ecaaea9873905b37658b9227c8fe99, and I will try to fix it tomorrow.
For this PR, I think you only need to focus on the semantic feature itself.

Reply:
Ok, I can revert the changes in `proto` and `provider` regarding that and commit; I'm testing them out in my own VSCode instance and they prevent flickering, so something along those lines is probably the way to go